src/submodules/fdtests.jl

```diff
@@ -0,0 +1,199 @@
+
+
+dirfdtest(F, dF, x, u; kwargs...) =
+      fdtest(t -> F(x + t * u),
+             t -> dot(dF(x + t * u), u),
+             0.0; kwargs...)
+
+
+
+# -------------------------------------------
+#  Interface code to perform FD tests on 
+#  systems with calculators 
+
+function _rattle(X, bb, r)
+   if r === false; return X, bb; end
+   X1 = copy(X)
+   TX = eltype(ustrip.(X)) 
+   bb1 = [bb...]
+   for i = 1:length(X)
+      ui = randn(TX); ui /= norm(ui)
+      X1[i] += (rand() * r) * ui
+   end
+   for i = 1:length(bb) 
+      ui = randn(TX); ui /= norm(ui)
+      bb1[i] += (rand() * r) * ui
+   end
+   return X1, bb1
+end
+
+
+function _fdtest_forces(calc, sys::AbstractSystem, verbose, rattle) 
+   X0, bb0 = _rattle(position(sys), bounding_box(sys), rattle)
+
+   _at(X) = FastSystem(bb0, 
+                       boundary_conditions(sys), 
+                       X, 
+                       atomic_symbol(sys), 
+                       atomic_number(sys), 
+                       atomic_mass(sys))
+
+   F = X -> potential_energy(_at(X), calc)
+   dF = X -> - forces(_at(X), calc)
+
+   verbose && println("Forces finite-difference test")
+   f_result = fdtest(F, dF, X0; verbose = verbose )
+   return f_result 
+end
+
+
+function _fdtest_virial(calc, sys::AbstractSystem, verbose, rattle)
+   X0, C0 = _rattle(position(sys), bounding_box(sys), rattle)
+
+   # reference deformation is just the identify
+   F0 = [1.0 0.0 0.0 ; 0.0 1.0 0.0 ; 0.0 0.0 1.0]
+   f0 = F0[:] 
+
+   # this implements a system where cell and positions are deformed according 
+   # to the deformation matrix F obtained from the vector fvec 
+   function _atv(fvec) 
+      F = reshape(fvec, (3,3))
+      bb = Ref(F) .* C0 # transform the cell 
+      X = Ref(F) .* X0 # transform the positions
+      return FastSystem(bb, 
+                        boundary_conditions(sys), 
+                        X, 
+                        atomic_symbol(sys), 
+                        atomic_number(sys), 
+                        atomic_mass(sys))
+   end
+
+   F = fvec ->  ustrip(potential_energy(_atv(fvec), calc))
+   dF = fvec -> Vector( ( - ustrip.(virial(_atv(fvec), calc)) )[:] )
+
+   verbose && println("Virial finite-difference test")
+   v_result = fdtest(F, dF, f0; verbose = verbose )
+   return v_result 
+end
+
+
+"""
+```julia
+fdtest(calc, sys::AbstractSystem; kwargs...)
+```
+Performs a finite-difference test for a calculator on an atom system and 
+returns a named tuple with the results.
+
+### kwargs
+- `verbose=true` : print the results of the test
+- `rattle=false` : apply a random perturbation to the system before testing (to turn this on, set the amount of rattling, not `true`)
+- `test_virial=true` : test the virial
+- `test_forces=true` : test the forces
+"""
+function fdtest(calc, sys::AbstractSystem;
+                verbose = true, 
+                rattle = false, 
+                test_virial = true, 
+                test_forces = true, 
+                )
```

Comment:
AtomsCalculators uses `(system, calculator)` order should we use that here too?

Also, would it be good to have error tolerance as a parameter?

Reply:
- I reversed it since I thought it was the calculator being tested ie the calculator is the main input. But I don't feel strongly about it.
- the tolerance : I'll have to think about it. I would like something that will work with F64 machine precision codes for now so we can integrate FD tests in EP, and then have a second PR that addresses calculators with numerically inconsistent forces